### PR TITLE
[ENH] Implement dropdowns for filtering by pipeline status to replace native datatable filter UI

### LINKED ISSUE #57: [ENH] Implement dropdowns for pipeline status filters to enable functional "AND" behaviour with session filtering
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Currently, because the session filters made by dropdown selections are applied to the uploaded dataframe ('memory' data) as opposed to the active datatable data ('virtual' data), this means that the built-in datatable filtering always ends up happening **after** any session filtering, as opposed to both being applied simultaneously. This becomes an issue when using "AND"-type session queries. 

e.g., If you (i) select sessions "1" and "2" in the session dropdown with the "AND" operator ("I want to filter for session 1 and session 2 data for subjects who have _both_ sessions available") **and** (ii) have entered a query string of `"SUCCESS"` in the datatable filter for the `fmriprep-20.2.7` column ("I want to see rows with an fmriprep completion status of `"SUCCESS"`"), the resulting rows will be _any_ subject-sessions with a `"SUCCESS"` status _among those_ belonging to subjects with both session 1 and 2.

In other words, it is not currently possible to apply a query of: "Show me subjects who have an fmriprep completion status of `"SUCCESS"` for **both** session 1 and 2"

### New feature
To simplify the logic of how session and pipeline status filters are applied and also (hopefully) make things a bit more intuitive for the user, switch to having dynamically-generated dropdown filters for each pipeline instead of using the datatable filter UI for these columns.

### Steps to implement

~- [ ] change logic of dropdown session filter to operate on virtual data instead of memory~
  - This gets very complicated because the state of the virtual data is not fixed, so when you e.g. select "1" and then "3", the selection of "3" would operate on the result of having applied "1" already
- [x] disable datatable column filter UI for pipeline + session columns
- [x] implement dropdown filter for each available pipeline
- [x] only support single-value selections for pipeline dropdowns
- [x] create version of matching records plot to display when there are 0 matches in datatable
- [X] hide session dropdown filters when there is no uploaded data

### PR DESCRIPTION
**Pull request by alyssadai**
Closes #57 

Changes proposed in this pull request:

- Dynamic generation of pipeline-specific dropdowns to enable filtering by completion statuses in datatable
  - Datatable filters for `session` and pipeline-specific columns now disabled to avoid confusion
- Callback + utility functions for updating datatable as result of backend filtering updated to handle both session queries and status queries from dropdowns according to the selected operator (`"AND"` / `"OR"`) 
- Dependency updates: 
    - `dash` version bump to use pattern-matching callback selectors
    - `dash-testing-stub` added to reqs - seems to be needed for webdriver config during tests to work properly with current pkg versions
- A properly-styled version of the status bar plot of matching records is now displayed even when there are 0 matches
- Some updates to function names and user-facing text for clarity

## Checklist

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[BUG]`, `[DOC]`, `[INFRA]`, `[MAINT]`)
- [x] PR links to Github issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Code is properly formatted

### CODE REVIEW

File: proc_dash/app.py [hunk 1]
```diff
@@ -302,11 +311,14 @@ def toggle_dataset_name_dialog(
     return is_open, None, None
 
 
+# TODO: Refactor session related operations into separate callback that relies on memory-overview component
```
Comment (surchs) [inline proc_dash/app.py:314]: :+1:

File: proc_dash/app.py [hunk 2]
```diff
@@ -324,66 +336,142 @@ def process_bagel(contents, filename):
...
         "csv",
         dataset_summary,
         {"display": "block"},
     )
 
 
+@app.callback(
+    [
+        Output("pipeline-dropdown-container", "children"),
+        Output("interactive-datatable", "style_filter_conditional"),
```
Comment (surchs) [inline proc_dash/app.py:393]: why is this an output of a callback? Couldn't you just statically set this where the data table is defined? Unless I misunderstand, there won't be any condition where you now want the data table filter functionality enabled. So setting it statically would make more sense I think

File: proc_dash/app.py [hunk 3]
```diff
@@ -324,66 +336,142 @@ def process_bagel(contents, filename):
...
+    pipelines_dict,
+):
     if parsed_data is None:
         return None, None
 
     data = pd.DataFrame.from_dict(parsed_data)
 
-    if session_values:
-        data = util.filter_by_sessions(
+    if session_values or not all(v is None for v in status_values):
```
Comment (surchs) [inline proc_dash/app.py:466]: :cherries: but I think 
```suggestion
    if session_values or any(v is not None for v in status_values):
```
is clearer

File: proc_dash/app.py [hunk 4]
```diff
@@ -324,66 +336,142 @@ def process_bagel(contents, filename):
...
+        pipeline_selected_filters = dict(
+            zip(pipelines_dict.keys(), status_values)
```
Comment (surchs) [inline proc_dash/app.py:468]: just checking here: how are you sure that both zipped objects have the same ordering?

File: proc_dash/app.py [hunk 5]
```diff
@@ -474,18 +562,25 @@ def generate_overview_status_fig_for_participants(parsed_data):
...
-    statuses aggregated by pipeline. Counts of statuses in plot thus correspond to unique records (unique
-    participant-session combinations).
+    but including custom component filtering), generate stacked bar plot of pipeline_complete statuses aggregated
+    by pipeline. Counts of statuses in plot thus correspond to unique records (unique participant-session
+    combinations).
     """
     if data is not None:
-        return plot.plot_pipeline_status_by_records(
-            pd.DataFrame.from_dict(data)
+        data_df = pd.DataFrame.from_dict(data)
```
Comment (surchs) [inline proc_dash/app.py:576]: I think this is getting a bit hard to read. :cherries:: maybe using the full if-else syntax here would make it visually clearer how you are using all of these return routes

File: proc_dash/plotting.py [hunk 6]
```diff
@@ -64,27 +67,46 @@ def plot_pipeline_status_by_participants(data: pd.DataFrame):
...
         .groupby(["pipeline_name", "pipeline_complete"])
         .size()
         .reset_index(name="records")
     )
 
+    return fig_pipeline_status_by_records(status_counts)
+
+
+def plot_empty_pipeline_status_by_records(pipelines: list, statuses: list):
+    """Generates version of matching records plot for 0 matching records (plot contains no data but correct labels)"""
```
Comment (surchs) [inline proc_dash/plotting.py:80]: I don't find this docstring very clear tbh. If I understand correctly, you are making a dataframe with all possible combinations of pipeline names and completion statuses, and then set the number of records for each of these to 0. The fact that this also makes a plot is not as relevant to understanding the function. And I do think it should not make a plot. Rather `populate_empty_pipeline_completion` as a function makes more sense to me, and then you can feed the output to your plotter wherever you call this function from.

File: proc_dash/app.py [hunk 7]
```diff
@@ -474,18 +562,25 @@ def generate_overview_status_fig_for_participants(parsed_data):
...
+        if not data_df.empty:
+            return plot.plot_pipeline_status_by_records(data_df), {
+                "display": "block"
+            }
+        return plot.plot_empty_pipeline_status_by_records(
```
Comment (surchs) [inline proc_dash/app.py:581]: as mentioned in the function: I think this function should not return a figure itself, but rather the data you then feed to `plot.plot_pipeline_status_by_records`. That might also clean up your if-else forest a little bit here.

File: requirements.txt
Comment (surchs) [inline requirements.txt:1]: Now that we have version bumping bots, let's have these bumps go in with their own PR. Makes reasoning about history of version change easier, and keeps the diff of functional PRs cleaner.

**Review by surchs — APPROVED** [review]
Looks good @alyssadai. I gave it a read and some cursory functional testing. I think all changes are good to go :cook: 

Meta-comment about PRs: This PR was challenging to review - not just because I'm not as familiar with the codebase as you, but rather because of the length and multitude of changes. Here are a couple of things we can do to improve readability and hopefully speed up review:

- have separate PRs for separate things. Don't mix functional changes with UI changes. And don't mix internal refactors with the other two. And make a separate PR for version bumps (or just use the bots PRs for that)
- unless the changes are trivial: make separate PRs for unrelated functional changes. I am not sure if that is the case here, but I think there is at least a mix of drop-down functionality and some changes to the schema that are mixed together here

lmk what you think. These are not specific to this repo, I think we should do that on all repos

File: requirements.txt
Comment (alyssadai) [reply requirements.txt:1]: In this case the dependency updates were necessary for the functional changes (pattern-matching callback selectors only available with `dash >= 2.9`; `dash-testing-stub` because e2e webdriver-dependent tests were failing after this upgrade) 😅. 

I'll aim to make the PR description clearer next time (or just comment on the file directly), and will be sure to separate out version bumps that are unrelated to the functional changes 🙂

File: proc_dash/app.py (hunk 2 above)
Comment (alyssadai) [reply proc_dash/app.py:393]: It's because I'm only disabling the filter functionality for columns that describe specific pipelines, e.g. `fmriprep-20.2.7` and `freesurfer-6.0.1`, aka cols that I have implemented dropdowns + backend filtering logic for. So currently there are subject identifier columns like `participant_id`, `bids_id` where users can still use the built-in filtering for if they are looking for the info of a specific subject, for example (i.e., "simple" within-column filtering).

The reason why this disabling is done dynamically as opposed to where the datatable is defined in the `layout` is just because I don't know which pipelines are available in a given dataset/datatable (and therefore which cols to disable) until after the data is uploaded + parsed.

I think one alternative in order to be able to set this statically in the layout definition would be to define the list of columns that I _don't_ want to disable, and basically make the condition something like `if not in [list_of_cols]`. The issue is that this list may need to be updated whenever I implement a new column in the overview table.

Any thoughts on this @surchs?

Comment (surchs) [reply proc_dash/app.py:393]: oh no, I think that's a good reason. I just didn't catch that!
